consensus/tests/integration_test.rs

```diff
@@ -0,0 +1,338 @@
+        once(&server_config).chain(&other_configs).collect(),
+        voting_powers,
+    );
+
+    let mut server_node = Consensus::new(
+        create_test_dms(
+            server_config.clone(),
+            dms_key.clone(),
+            SharedKnownPeers::new_static(vec![]),
+        )
+        .await,
+        create_storage(create_temp_dir()).await,
+        block_header.clone(),
+        params.clone(),
+        round_zero_timestamp,
+        Some(server_config.private_key.clone()),
+    )
+    .await
+    .unwrap();
+    let mut other_nodes = Vec::new();
+    for config in &other_configs {
+        let consensus = Consensus::new(
+            create_test_dms(config.clone(), dms_key.clone(), peers.clone()).await,
+            create_storage(create_temp_dir()).await,
+            block_header.clone(),
+            params.clone(),
+            round_zero_timestamp,
+            Some(config.private_key.clone()),
+        )
+        .await
+        .unwrap();
+        other_nodes.push(consensus);
+    }
+
+    // Make a block to propose
+    let dummy_block_hash = Hash256::hash("dummy_block");
+    server_node
+        .register_verified_block_hash(dummy_block_hash)
+        .await
+        .unwrap();
+    for other_node in &mut other_nodes {
+        other_node
+            .register_verified_block_hash(dummy_block_hash)
+            .await
+            .unwrap();
+    }
+
+    // [Step 1]
+    // Action: The server node proposes a block
+    // Expected: The server node will propose a block and prevotes on it.
+    // Initial block candidate is set to 0 by default, so we progress without setting a new candidate.
+    let timestamp = get_timestamp();
+    let mut result = Vec::new();
+    result.extend(
+        server_node
+            .set_proposal_candidate(dummy_block_hash, timestamp)
+            .await
+            .unwrap(),
+    );
+    result.extend(server_node.progress(get_timestamp()).await.unwrap());
+    let expected = vec![
+        ProgressResult::Proposed(0, dummy_block_hash, timestamp),
+        ProgressResult::NonNilPreVoted(0, dummy_block_hash, timestamp),
+    ];
+    assert_eq!(result, expected);
+
+    // Action: Non-server nodes fetch messages from the server and make progress.
+    // Expected: Node 0, 1 will prevote, node 2, 3 will precommit.
+    let serve_task = tokio::spawn(async { server_node.serve(3_000).await });
+    for (i, other_node) in other_nodes.iter_mut().enumerate() {
+        println!("Checking node #{}", i);
+        other_node.fetch().await.unwrap();
+        let timestamp = get_timestamp();
+        let result = other_node.progress(timestamp).await.unwrap();
+        let mut expected = vec![ProgressResult::NonNilPreVoted(
+            0,
+            dummy_block_hash,
+            timestamp,
+        )];
+        // The nodes will broadcast precommits as well if they see prevotes over 2/3.
+        // Plus-ones for the server and the node itself.
+        if i == 2 || i == 3 {
```

Comment:
We need to sleep before this, to ensure that the server node served the messages well. 500ms will be enough.